pages/en/lb4/Testing-Your-application.md

```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
+
+```ts
+describe('ProductController', () => {
+  let repository: ProductRepository;
+  beforeEach(givenStubbedRepository);
+
+  // your unit tests
+
+  function givenStubbedRepository() {
+    repository = sinon.createStubInstance(ProductRepository);
+  }
+});
+```
+
+In your unit-tests, you will usually want to program the behavior of stubbed methods (what should they return) and then verify that the Controller (unit under test) called the right method with the correct arguments.
+
+Configure stub's behavior at the beginning of your unit-test (in the "arrange" or "given" section):
+
+```ts
+// repository.find() will return a promise that
+// will be resolved with the provided array
+const findStub = repository.find as sinon.SinonStub;
+findStub.resolves([{id: 1, name: 'Pen'}]);
+```
+
+Verify how was the stubbed method executed at the end of your unit-test (in the "assert" or "then" section):
+
+```ts
+// expect that repository.find() was called with the first
+// argument deeply-equal to the provided object
+expect(findStub).to.be.calledWithMatch({where: {id: 1}});
+```
+
+See [Unit test your controllers](#unit-test-your-controllers) for a full example.
+
+#### Create a stub Service
+
+To be done. The initial Beta release does not include Services as a first-class feature.
+
+### Unit-test your Controllers
+
+A typical unit-test creates a controller instance with dependencies replaced by Test doubles and directly calls the tested method. In the example below, we give the controller a stub implementation of its repository dependency, see [Create a stub repository](#create-a-stub-repository) for a detailed explanation.
+
+```ts
+// test/controllers/product.controller.unit.ts
+import {ProductController, ProductRepository} from '../..';
+import {expect, sinon} from '@loopback/testlab';
+
+describe('ProductController (unit)', () => {
+  let repository: ProductRepository;
+  beforeEach(givenStubbedRepository);
+
+  describe('getDetails()', () => {
+    it('retrieves details of a product', async () => {
+      const controller = new ProductController(repository);
+      const findStub = repository.find as sinon.SinonStub;
+      findStub.resolves([{id: 1, name: 'Pen'}]);
+
+      const details = await controller.getDetails(1);
+
+      expect(details).to.containDeep({name: 'Pen'});
+      expect(findStub).to.be.calledWithMatch({where: {id: 1}});
+    });
+  });
+
+  function givenStubbedRepository() {
+    repository = sinon.createStubInstance(ProductRepository);
+  }
+});
+```
+
+### Unit-test your Models and Repositories
+
+In a typical LoopBack appliction, models and repositories are relying mostly on the behavior provided by the framework (`@loopback/repository` package) and there is no need to test the LoopBack's built-in functionality unit-test level (the functionality has been already verified by the tests in the framework). What does need new unit-tests is any additional application-specific API.
```

Comment:
Break up into multiple lines.

Reply:
`appliction` -> `application`

Reply:
`...and there is no need to test the LoopBack's built-in functionality unit-test level...`
Awkward phrasing. I'd probably rephrase it to be direct about not needing to unit test anything we provide:
```
Your unit tests should not attempt to validate classes or functions provided by the
`@loopback/repository` (ex. EntityCrudRepository).
Instead, use test doubles to stub or mock these classes and methods.
```